src/Storages/NATS/ReadBufferFromNATSConsumer.cpp

```diff
@@ -0,0 +1,119 @@
+#include <atomic>
+#include <chrono>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <utility>
+#include <Storages/NATS/ReadBufferFromNATSConsumer.h>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/bind.hpp>
+#include "Poco/Timer.h"
+#include <Common/logger_useful.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int CANNOT_CONNECT_NATS;
+}
+
+ReadBufferFromNATSConsumer::ReadBufferFromNATSConsumer(
+    std::shared_ptr<NATSConnectionManager> connection_,
+    std::vector<String> & subjects_,
+    const String & subscribe_queue_name,
+    Poco::Logger * log_,
+    char row_delimiter_,
+    uint32_t queue_size_,
+    const std::atomic<bool> & stopped_)
+    : ReadBuffer(nullptr, 0)
+    , connection(connection_)
+    , subjects(subjects_)
+    , log(log_)
+    , row_delimiter(row_delimiter_)
+    , stopped(stopped_)
+    , queue_name(subscribe_queue_name)
+    , received(queue_size_)
+{
+}
+
+ReadBufferFromNATSConsumer::~ReadBufferFromNATSConsumer()
+{
+    for (const auto & subscription : subscriptions)
+    {
+        natsSubscription_Unsubscribe(subscription.get());
+    }
```

Comment:
just call `unsubscribe()`?

Reply:
I don't event need it here really, because when unique ptr on subscription will get destroyed, it will call natsSubscription_Destroy which will automatically unsubscribe. So with the next line being redundant, I think we can just make the destructor default.